### PR TITLE
ESAPI Change the default TCTI initialization

### PR DESCRIPTION
**Pull request by AndreasFuchsSIT**
Two very minor commit first that I just split out before the actual changes.

The ESAPI gains a new behavior for initializing the
default TCTI.

- The configure flag is now interpreted as a file
  instead of as a partial symbol name.
- If a default tcti is configured only this file is
  loaded.
- If no default tcti is configured, esys will attempt
  to load a tcti in the following order
  - Search for a libtss2-tcti-default.so
  - Search for a libtss2-tcti-tabrmd.so
  - Attempt the linked Tcti_Device for "/dev/tpmrm0"
  - Attempt the linked Tcti_Device for "/dev/tpm0"
  - Attempt the linked Tcti_Mssim for "localhost:2321"

@williamcroberts I'd like to get your opinion from the tpm2-tools perspective as well.

### CODE REVIEW

File: src/tss2-esys/esys_tcti_default.c [hunk 1]
```diff
@@ -1,56 +1,214 @@
...
+
+    handle = dlopen(file, RTLD_NOW);
+    if (handle == NULL) {
+        LOG_WARNING("Could not load TCTI file: %s", file);
+        return TSS2_ESYS_RC_BAD_REFERENCE;
+    }
+
+    infof = (TSS2_TCTI_INFO_FUNC) dlsym(handle, TSS2_TCTI_INFO_SYMBOL);
+    if (infof == NULL) {
+        LOG_ERROR("Info not found in TCTI file: %s", file);
```
Comment (tstruk) [inline src/tss2-esys/esys_tcti_default.c]: dlclose()?

File: src/tss2-esys/esys_tcti_default.c [hunk 2]
```diff
@@ -1,56 +1,214 @@
...
+        LOG_ERROR("tcticontext must not be NULL");
+        return TSS2_TCTI_RC_BAD_REFERENCE;
+    }
+    *tcticontext = NULL;
+#ifdef ESYS_TCTI_DEFAULT_MODULE
+
+#ifdef ESYS_TCTI_DEFAULT_CONFIG
+    const char *config = _XSTR(ESYS_TCTI_DEFAULT_CONFIG);
+#else /* ESYS_TCTI_DEFAULT_CONFIG */
+    const char *config = "";
```
Comment (tstruk) [inline src/tss2-esys/esys_tcti_default.c]: should be const char *config = NULL;

File: src/tss2-esys/esys_tcti_default.c [hunk 3]
```diff
@@ -1,56 +1,214 @@
...
+#endif /* ESYS_TCTI_DEFAULT_CONFIG */
+
+    LOG_DEBUG("Attempting to initialize TCTI defined during compilation: %s:%s",
+              ESYS_TCTI_DEFAULT_MODULE, config);
+    return tcti_from_file(ESYS_TCTI_DEFAULT_MODULE, config, tcticontext);
+
+#else /* ESYS_TCTI_DEFAULT_MODULE */
```
Comment (tstruk) [inline src/tss2-esys/esys_tcti_default.c]: Is this else needed. If ESYS_TCTI_DEFAULT_MODULE is defined we call tcti_from_file(ESYS_TCTI_DEFAULT_MODULE,..) and return.

File: src/tss2-esys/esys_tcti_default.c [hunk 4]
```diff
@@ -1,56 +1,214 @@
...
- * Initialize a TCTI instance.
- * The caller is returned a TCTI context structure that is allocated by this
- * function. This structure must be freed by the caller.
- */
-TSS2_RC
-get_tcti_default(TSS2_TCTI_CONTEXT ** tcticontext)
+#define ARRAY_SIZE(X) (sizeof(X)/sizeof(X[0]))
+
+struct {
+    char *file;
```
Comment (tstruk) [inline src/tss2-esys/esys_tcti_default.c]: #ifndef NO_DL around file?

Comment (AndreasFuchsSIT) [reply src/tss2-esys/esys_tcti_default.c]: @tstruk I think I don't know what you mean ?

File: src/tss2-esys/esys_tcti_default.c (hunk 3 above)
Comment (AndreasFuchsSIT) [reply src/tss2-esys/esys_tcti_default.c]: Exactly. Thus code-scanners could complain about unreachable code after return. Thus the #else...

File: src/tss2-esys/esys_tcti_default.c [hunk 5]
```diff
@@ -1,56 +1,214 @@
...
+        } else {
+            LOG_ERROR("Errorous entry in standard TCTIs");
+            return TSS2_ESYS_RC_GENERAL_FAILURE;
+        }
+
+    }
+#endif /* ESYS_TCTI_DEFAULT_MODULE */
+
+    LOG_ERROR("No standard TCTI could be loaded");
+    return TSS2_ESYS_RC_NOT_IMPLEMENTED;
```
Comment (AndreasFuchsSIT) [inline src/tss2-esys/esys_tcti_default.c]: But I guess this needs to go before the #endif...

File: src/tss2-esys/esys_tcti_default.c (hunk 1 above)
Comment (AndreasFuchsSIT) [reply src/tss2-esys/esys_tcti_default.c]: Done

File: src/tss2-esys/esys_tcti_default.c (hunk 2 above)
Comment (AndreasFuchsSIT) [reply src/tss2-esys/esys_tcti_default.c]: done

File: src/tss2-esys/esys_tcti_default.c (hunk 4 above)
Comment (tstruk) [reply src/tss2-esys/esys_tcti_default.c]: Just saying that we could also wrap the char *file member with #ifndef NO_DL:
```
struct {
    #ifndef NO_DL
    char *file;
    #endif
    TSS2_TCTI_INIT_FUNC init;
    TSS2_RC	+    char *conf;
    get_tcti_default(TSS2_TCTI_CONTEXT ** tcticontext)	+    char *description;
};
```

Comment (AndreasFuchsSIT) [reply src/tss2-esys/esys_tcti_default.c]: ok, done...

**Comment by AndreasFuchsSIT** [comment]
@tstruk hope this is ok now...

**Pull request merged**